resource-managers/kubernetes/token-refresh-server/src/main/scala/org/apache/spark/security/kubernetes/TokenRefreshService.scala

```diff
@@ -0,0 +1,361 @@
+      refreshService ! Renew(nextExpireTime, newExpireTimeByToken, renew.secretMeta,
+        numConsecutiveErrors)
+    } else {
+      logWarning(s"Got an empty token list with ${renew.secretMeta.getSelfLink}")
+    }
+  }
+
+  private def refresh(token: Token[_ <: TokenIdentifier], expireTime: Long, deadline: Long,
+                      nowMillis: Long) = {
+    val maybeNewToken = maybeObtainNewToken(token, expireTime, nowMillis)
+    val maybeNewExpireTime = maybeGetNewExpireTime(maybeNewToken, expireTime, deadline, nowMillis)
+    (maybeNewToken, maybeNewExpireTime)
+  }
+
+  private def maybeObtainNewToken(token: Token[_ <: TokenIdentifier], expireTime: Long,
+                                  nowMills: Long) = {
+    val maybeNewToken = if (token.getKind.equals(DelegationTokenIdentifier.HDFS_DELEGATION_KIND)) {
+      // The token can casted to AbstractDelegationTokenIdentifier below only if the token kind
+      // is HDFS_DELEGATION_KIND, according to the YARN resource manager code. See if this can be
+      // generalized beyond HDFS tokens.
+      val identifier = token.decodeIdentifier().asInstanceOf[AbstractDelegationTokenIdentifier]
+      val maxDate = identifier.getMaxDate
+      if (maxDate - expireTime < RENEW_TASK_REMAINING_TIME_BEFORE_NEW_TOKEN_MILLIS ||
+        maxDate <= nowMills) {
+        logDebug(s"Obtaining a new token with maxData $maxDate," +
+          s" expireTime $expireTime, now $nowMills")
+        val newToken = obtainNewToken(token, identifier)
+        logInfo(s"Obtained token $newToken")
+        newToken
+      } else {
+        token
+      }
+    } else {
+      token
+    }
+    maybeNewToken
+  }
+
+  private def maybeGetNewExpireTime(token: Token[_ <: TokenIdentifier], expireTime: Long,
+                                    deadline: Long,
+                                    nowMillis: Long) = {
+    if (expireTime <= deadline || expireTime <= nowMillis) {
+      try {
+        logDebug(s"Renewing token $token with current expire time $expireTime," +
+          s" deadline $deadline, now $nowMillis")
+        val newExpireTime = token.renew(hadoopConf)
+        logDebug(s"Renewed token $token. Next expire time $newExpireTime")
+        newExpireTime
+      } catch {
+        case t: Throwable =>
+          logWarning(t.getMessage, t)
+          hasError = true
+
+          getRetryTime
+      }
+    } else {
+      expireTime
+    }
+  }
+
+  private def obtainNewToken(token: Token[_ <: TokenIdentifier],
+                             identifier: AbstractDelegationTokenIdentifier) = {
+    val owner = identifier.getOwner
+    val realUser = identifier.getRealUser
+    val user = if (realUser == null || realUser.toString.isEmpty || realUser.equals(owner)) {
+      owner.toString
+    } else {
+      realUser.toString
+    }
+    val credentials = new Credentials()
+    val ugi = UserGroupInformation.createProxyUser(user, UserGroupInformation.getLoginUser)
```

Comment:
Why create a proxy user instead of just using `UGI.getLoginUser`?

Reply:
The refresh server is designed to serve multiple Spark jobs. These jobs may come from many different users. The refresh server has its own headless Kerberos account as the login user, which is `UGI.getLoginUser`. On the other hand, the HDFS tokens it creates belong to the job user accounts because the job's driver and executors should use them to access the job user data. The proxy user allows the refresh server to use its own account to obtain job user tokens. If it uses the login user, the obtained tokens would not be able to access the job user data.